src/ui/AuthenticationProvider/AuthenticationProvider.ts

```diff
@@ -136,6 +142,51 @@ export class AuthenticationProvider extends Component {
     });
   }
 
+  private getHandshakeTokenFromUrl() {
+    const fragment = window.location.hash.slice(1);
+    const params = fragment.split('&');
+
+    const tokenParam = _.find(params, param => {
+      const [key] = param.split('=');
+      return key === 'handshake_token';
+    });
+
+    if (!tokenParam) {
+      return '';
+    }
+
+    const token = tokenParam.split('=')[1];
+    return token ? decodeURIComponent(token) : '';
+  }
+
+  private onAfterComponentsInitialization(args: IInitializationEventArgs) {
+    const handshakeToken = this.getHandshakeTokenFromUrl();
+
+    if (!handshakeToken) {
+      return this.loadAccessTokenFromStorage();
+    }
+
+    const promise = this.exchangeHandshakeToken(handshakeToken)
+      .then(token => this.storeAccessToken(token))
+      .then(() => this.loadAccessTokenFromStorage())
+      .catch(e => this.logger.error(e));
+
+    args.defer.push(promise);
+  }
+
+  private exchangeHandshakeToken(token: string) {
+    return SearchEndpoint.defaultEndpoint.exchangeAuthenticationProviderHandshakeTokenForAccessToken(token);
```

Comment:
We should not assume `.defaultEndpoint`. 

```suggestion
    return this.queryController.getEndpoint().exchangeAuthenticationProviderHandshakeTokenForAccessToken(token);
```